transfer-sh.el

```diff
@@ -54,39 +54,76 @@
+If no REMOTE-FILENAME is provided, read from minibuffer the
+remote file name visible in the transfer.sh link and run
+`transfer-sh-run-upload-agent'."
   (interactive "ffile: ")
+  (or remote-filename
+      (setq remote-filename (read-from-minibuffer
+                             (format "Remote filename (default %s): "
+                                     (file-name-nondirectory local-filename))
+                             (file-name-nondirectory local-filename))))
   (async-start
    `(lambda ()
       ,(async-inject-variables "local-filename")
       ,(async-inject-variables "remote-filename")
-      (shell-command-to-string
-        (concat "curl --silent --upload-file "
-                (shell-quote-argument local-filename)
-                " " (shell-quote-argument (concat "https://transfer.sh/" remote-filename)))))
-   `(lambda (transfer-link)
-      (kill-new transfer-link)
-      (message transfer-link))))
+      ,(transfer-sh-run-upload-agent local-filename remote-filename))))
 
 ;;;###autoload
 (defun transfer-sh-upload-file (local-filename &optional remote-filename)
   "Uploads file LOCAL-FILENAME to transfer.sh.
 
-If no REMOTE-FILENAME is given, the LOCAL-FILENAME is used."
+If no REMOTE-FILENAME is provided, read from minibuffer the
+remote file name visible in the transfer.sh link and run
+`transfer-sh-run-upload-agent'."
   (interactive "ffile: ")
-  (let*  ((remote-filename (if remote-filename
-                               remote-filename
-                             (file-name-nondirectory local-filename)))
-          (transfer-link (shell-command-to-string
-                           (concat "curl --silent --upload-file "
-                                   (shell-quote-argument local-filename)
-                                   " " (shell-quote-argument (concat "https://transfer.sh/" remote-filename))))))
+  (transfer-sh-run-upload-agent
+   local-filename
+   (or remote-filename
+       (read-from-minibuffer
+        (format "Remote filename (default %s): "
+                (file-name-nondirectory local-filename))
+        (file-name-nondirectory local-filename)))))
+
+(defun transfer-sh-run-upload-agent (local-filename  &optional remote-filename)
+  "Uploads LOCAL-FILENAME to transfer.sh using `transfer-sh-upload-agent-command'.
+
+If no REMOTE-FILE is given, LOCAL-FILENAME is used."
+  (let* ((filename-without-directory (file-name-nondirectory local-filename))
+         (remote-filename (or remote-filename filename-without-directory))
+         (transfer-link (with-temp-buffer
+                          (apply 'call-process
+                                 transfer-sh-upload-agent-command
+                                 nil t nil
+                                 (append transfer-sh-upload-agent-arguments
+                                         (list local-filename
+                                               (concat "https://transfer.sh/" remote-filename))))
```

Comment:
If I'm not mistaken, this won't work if remote-file contains some characters like spaces.
If it does not, enclose it in `url-encode-url`.

Reply:
Very good idea, thanks!